documentation/specs/proposed/BuildCheck.md

```diff
@@ -0,0 +1,396 @@
+* `.editorconfig` files placed along with explicitly or implicitly imported msbuild files won’t be considered.
+* `.editorconfig` files packaged within nuget packages within local nuget cache won’t be considered.
+
+### Non-Goals (but might be considered):
+* bulk configuration of multiple rules - based on analyzers/rules prefixes or/and categories.
+* attempts to try to configure standard msbuild warnings/errors via `.editorconfig` should lead to fail fast errors.
+* configuring analysis levels when analysing from binlog - beyond the collected editorconfigs.
+* Aliasing the analyzers/rules, allowing to create multiple instances with different custom configuration (e.g. single analyzer checking configurable list of forbidden properties prefixes can have 2 instance, each initialized with different list to check, each of the instance configurable for individual projects separately).
+
+### Out of scope for configuration:
+* opt-out of analysis on code-level (analogy to C# pragmas, but within msbuild xml files).
+* lower granularity of `.editorconfig` settings other than whole projects.
+* attributing configuration to a .sln file and expecting it will apply to all contained projects.
+* Support for multiple [custom configurations](#custom-configuration-declaration) within a single build for a single rule. (Not to be mixed with [standardized configuration](#standardized-configuration-declaration) - which can be configured freely per project) If a custom configuration will be used, it will need to be specified identically in each explicit configurations of the rule. This is chosen so that there are no implicit requirements on lifetime of the analyzer or analyzer instancing – each analyzer will be instantiated only once per build (this is however something that will very likely change in future versions – so authors are advised not to take hard dependency on single instance policy).
+
+### Sample configuration
+
+```ini
+[*.csproj]
+msbuild_analyzer.BC0101.IsEnabled=true
+msbuild_analyzer.BC0101.Severity=warning
+
+msbuild_analyzer.COND0543.IsEnabled=false
+msbuild_analyzer.COND0543.Severity=Error
+msbuild_analyzer.COND0543.EvaluationAnalysisScope=AnalyzedProjectOnly
+msbuild_analyzer.COND0543.CustomSwitch=QWERTY
+```
+
+### User Configurable Options
+
+Initial version of BuildCheck plans a limited set of options configurable by user (via `.editorconfig`) by which users can override default configuration of individual analyzer rules.
+
+#### Enablement
+
+Boolean option `IsEnabled` will be available to allow users to disable/enable particualr rule.
+
+Different rules of a single analyzer can have different enablement status configured.
+
+If all the rules from a single analyzer are disabled - analyzer won't be given any data for such configured part of the build (specific project or a whoe build). If analyzer have some rules enabled and some disabled - it will be still fed with data, but the reports will be post-filtered.
```

Comment:
> **If analyzer have some rules enabled and some disabled ...**

This begs the question of whether the analyzer can know which of its rules are disabled, to give it the ability to avoid potential expensive processing.

Reply:
https://github.com/orgs/dotnet/projects/373?pane=issue&itemId=57761219